### Loading common data formats   
[**Neal Caren**](mailto:neal.caren@gmail.com)  
University of North Carolina, Chapel Hill

It is possible to turn many sorts of data into a Pandas dataframe for subsequent anaysis. 

The most basic method is reading comma-delimited text files, or csv files. This is accomplished with the `read_csv`.

In [7]:
import pandas as pd

In [8]:
df = pd.read_csv("data/titanic.csv")

After a loading any new data, it is best practice to look at the table and basic descriptive statistics. This both ensures that the data loaded correctly and gives you an overview of the data.

First, take a look at the top ten rows.

In [9]:
df.head(10)


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Note that pandas has created an index which is shown furtherest to the left. By default this is sequential integer counting up from zero. Alternatively, if you have a column or column that can be used to identify cases, you can assign them to be the index.

In [ ]:
df_indexed = pd.read_csv("data/titanic.csv", index_col="PassengerId")

df_indexed.head(10)


Meaninful indexes are particularly useful when you want to merge dataframes.

You can look at the bottom few rows to make sure the data format is relatively consistent. 

In [5]:
df.tail(5)

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

`info` details information about the dataframe such as number of rows and columns, along with information about each column, including the data type and the number of non-missing values.

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


`describe` shows the distribution for numerical columns.

In [13]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Pandas assumes that the file is comma delimited. If that is not the case, the `sep` parameter can be set to whatever the delimiter is. Most comman options would be `\t` for tab and `|` for pipe.

In [12]:
df_tab = pd.read_csv("data/titanic_tab.csv", sep="\t", index_col="PassengerId")

df_tab.head(5)


Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

Excel `xls` files can also be read in directly. You need to specify which sheet.

In [15]:
df_xls = pd.read_excel("data/titanic.xls", sheet_name="Sheet1")


In [16]:
df_xls.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

There are similar methods for reading SAS XPORT or SAS7BDAT files with `pd.read_SAS()` and Stata .dta files with `pd.read_dta()`. Unfortunately, `read_dta` does not work with the most current version of Stata files, which is also true of older versions of Stata.

### Beyond the Basics

JSON is increasingly common file format for social scientists to encounter. It is a human-readable text format where each record includes both the variable names, such as `Sex` or `Fare` along with values, like "male" or 7.25. When viewed in a text editor, the contents are often rendered in a fairly straightforward way.

<img src="images/titanic_json.png" width="800">

Importing a JSON file as a Pandas dataframe is similar to the other import processes. 

In [17]:
df_json = pd.read_json('data/titanic.json')

In [19]:
df_json.head()

Age Cabin Embarked     Fare  \
0  22.0  None        S   7.2500   
1  38.0   C85        C  71.2833   
2  26.0  None        S   7.9250   
3  35.0  C123        S  53.1000   
4  35.0  None        S   8.0500   

                                                Name  Parch  PassengerId  \
0                            Braund, Mr. Owen Harris      0            1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0            2   
2                             Heikkinen, Miss. Laina      0            3   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0            4   
4                           Allen, Mr. William Henry      0            5   

   Pclass     Sex  SibSp  Survived            Ticket  
0       3    male      1         0         A/5 21171  
1       1  female      1         1          PC 17599  
2       3  female      0         1  STON/O2. 3101282  
3       1  female      1         1            113803  
4       3    male      0         0            373450

If your data comes as an XML 